### PR TITLE
QAT anomaly example

### PR DESCRIPTION
**Pull request by alexsu52**
### Changes

Added the example: Quantization-Aware Training of STFPM PyTorch model from Anomalib

### Reason for changes

This example demonstrates how to quantize [Student-Teacher Feature Pyramid Matching (STFPM)](https://anomalib.readthedocs.io/en/latest/markdown/guides/reference/models/image/stfpm.html) PyTorch model from [Anomalib](https://github.com/openvinotoolkit/anomalib) using Quantization API from Neural Network Compression Framework (NNCF).

### Related tickets

N/A

### Tests

test_examples 323

### CODE REVIEW

File: examples/quantization_aware_training/torch/anomalib/main.py [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+    print("Test results for INT8 model after PTQ:")
+    int8_init_test_results = engine.test(model=quantized_model, datamodule=datamodule)
+
+    ###############################################################################
+    # Step 3: Fine tune the quantized model
+    print(os.linesep + "[Step 3] Fine tune the quantized model")
+
+    engine.fit(model=quantized_model, datamodule=datamodule)
+    print("Test results for INT8 model after QAT:")
+    int8_test_results = engine.test(model=quantized_model, datamodule=datamodule)
```
Comment (daniil-lyakhov) [inline examples/quantization_aware_training/torch/anomalib/main.py:117]: Please print result of validation here as it done in resnet18 example https://github.com/openvinotoolkit/nncf/blob/develop/examples/quantization_aware_training/torch/resnet18/main.py#L270

File: examples/quantization_aware_training/torch/anomalib/main.py [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+    # Step 6: Summary
+    print(os.linesep + "[Step 6] Summary")
+
+    fp32_f1score = fp32_test_results[0]["image_F1Score"]
+    int8_init_f1score = int8_init_test_results[0]["image_F1Score"]
+    int8_f1score = int8_test_results[0]["image_F1Score"]
+
+    print(f"Accuracy drop after PTQ:                {fp32_f1score - int8_init_f1score:.3f}")
+    print(f"Accuracy drop after QAT:                {fp32_f1score - int8_f1score:.3f}")
+    print(f"Model compression rate:                 {fp32_size / int8_size:.3f}")
```
Comment (daniil-lyakhov) [inline examples/quantization_aware_training/torch/anomalib/main.py:153]: What do you think about a summary table as it does in the resnet18 qat example?

File: tests/cross_fw/examples/example_scope.json [hunk 3]
```diff
@@ -207,5 +207,27 @@
...
+        "backend": "torch",
+        "requirements": "examples/quantization_aware_training/torch/anomalib/requirements.txt",
+        "cpu": "Intel(R) Core(TM) i9-10980XE CPU @ 3.00GHz",
+        "accuracy_tolerance": 0.02,
+        "accuracy_metrics": {
+            "fp32_f1score": 0.9919999837875366,
+            "int8_init_f1score": 0.9767441749572754,
+            "int8_f1score": 0.9919999837875366,
+            "accuracy_drop": 0.0
+        },
```
Comment (daniil-lyakhov) [inline tests/cross_fw/examples/example_scope.json:221]: Let's unify this test with the latest fix https://github.com/openvinotoolkit/nncf/pull/2594

File: examples/quantization_aware_training/torch/anomalib/README.md [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+
+The anomaly detection domain is one of the domains in which models are used in scenarios where the cost of model error is high and accuracy cannot be sacrificed for better model performance. Quantization-Aware Training (QAT) is perfect for such cases, as it reduces quantization error without model performance degradation by training the model.
+
+This example demonstrates how to quantize [Student-Teacher Feature Pyramid Matching (STFPM)](https://anomalib.readthedocs.io/en/latest/markdown/guides/reference/models/image/stfpm.html) PyTorch model from [Anomalib](https://github.com/openvinotoolkit/anomalib) using Quantization API from Neural Network Compression Framework (NNCF). At the first step, the model is quanitzed using Post-Training Quantization (PTQ) algorithm to obtain the best initialization of the quantized model. If the accuracy of the quantized model after PTQ does not meet requiremenets, the next step is to train the quantized model using PyTorch framework.
+
+NNCF provides semiless transition from Post-Training Quantization to Quantization-Aware Training without additional model preparation and transfer of magic parameters.
```
Comment (daniil-lyakhov) [inline examples/quantization_aware_training/torch/anomalib/README.md:7]:
```suggestion
NNCF provides seamless transition from Post-Training Quantization to Quantization-Aware Training without additional model preparation and transfer of magic parameters.
```

File: examples/quantization_aware_training/torch/anomalib/main.py (hunk 1 above)
Comment (alexsu52) [reply examples/quantization_aware_training/torch/anomalib/main.py:117]: `engine.test` prints results the following way: 
image

File: examples/quantization_aware_training/torch/anomalib/main.py (hunk 2 above)
Comment (alexsu52) [reply examples/quantization_aware_training/torch/anomalib/main.py:153]: Printing the table takes a lot of lines, I prefer a simple solution.

File: tests/cross_fw/examples/example_scope.json (hunk 3 above)
Comment (alexsu52) [reply tests/cross_fw/examples/example_scope.json]: done

**Review by daniil-lyakhov — APPROVED** [review]

File: examples/quantization_aware_training/torch/anomalib/README.md [hunk 5]
```diff
@@ -0,0 +1,37 @@
+# Quantization-Aware Training of STFPM PyTorch model from Anomalib
+
+The anomaly detection domain is one of the domains in which models are used in scenarios where the cost of model error is high and accuracy cannot be sacrificed for better model performance. Quantization-Aware Training (QAT) is perfect for such cases, as it reduces quantization error without model performance degradation by training the model.
+
+This example demonstrates how to quantize [Student-Teacher Feature Pyramid Matching (STFPM)](https://anomalib.readthedocs.io/en/latest/markdown/guides/reference/models/image/stfpm.html) PyTorch model from [Anomalib](https://github.com/openvinotoolkit/anomalib) using Quantization API from Neural Network Compression Framework (NNCF). At the first step, the model is quanitzed using Post-Training Quantization (PTQ) algorithm to obtain the best initialization of the quantized model. If the accuracy of the quantized model after PTQ does not meet requiremenets, the next step is to train the quantized model using PyTorch framework.
```
Comment (AlexKoff88) [inline examples/quantization_aware_training/torch/anomalib/README.md:5]:
```suggestion
This example demonstrates how to quantize [Student-Teacher Feature Pyramid Matching (STFPM)](https://anomalib.readthedocs.io/en/latest/markdown/guides/reference/models/image/stfpm.html) PyTorch model from [Anomalib](https://github.com/openvinotoolkit/anomalib) using Quantization API from Neural Network Compression Framework (NNCF). In the first step, the model is quantized using Post-Training Quantization (PTQ) algorithm to obtain the best initialization of the quantized model. If the accuracy of the quantized model after PTQ does not meet requirements, the next step is to train the quantized model using PyTorch framework.
```

File: examples/quantization_aware_training/torch/anomalib/README.md (hunk 4 above)
Comment (AlexKoff88) [inline examples/quantization_aware_training/torch/anomalib/README.md:7]:
```suggestion
NNCF provides a seamless transition from Post-Training Quantization to Quantization-Aware Training without additional model preparation and transfer of magic parameters.
```

File: examples/quantization_aware_training/torch/anomalib/README.md [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+
+NNCF provides semiless transition from Post-Training Quantization to Quantization-Aware Training without additional model preparation and transfer of magic parameters.
+
+The example includes the following steps:
+
+- Loading the [MVTec (capsule category)](https://www.mvtec.com/company/research/datasets/mvtec-ad) dataset (~4.9 Gb).
+- (Optional) Training STFPM PyTorch model from scratch.
+- Loading STFPM model pretrained on this dataset.
+- Quantizing the model using NNCF Post-Training Quantization algorithm.
+- Fine tuning quantized model for one epoch to improve quantized model metrics.
```
Comment (AlexKoff88) [inline examples/quantization_aware_training/torch/anomalib/README.md:15]:
```suggestion
- Fine-tuning quantized model for one epoch to improve quantized model metrics.
```

File: examples/quantization_aware_training/torch/anomalib/README.md [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+- Output of the following characteristics of the quantized model:
+  - Accuracy drop of the quantized model (INT8) over the pre-trained model (FP32)
+  - Compression rate of the quantized model file size relative to the pre-trained model file size
+  - Performance speed up of the quantized model (INT8)
+
+## Install requirements
+
+At this point it is assumed that you have already installed NNCF. You can find information on installation NNCF [here](https://github.com/openvinotoolkit/nncf#user-content-installation).
```
Comment (AlexKoff88) [inline examples/quantization_aware_training/torch/anomalib/README.md:23]:
```suggestion
At this point, it is assumed that you have already installed NNCF. You can find information on installation of NNCF [here](https://github.com/openvinotoolkit/nncf#user-content-installation).
```

File: examples/quantization_aware_training/torch/anomalib/main.py [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+
+import nncf
+
+HOME_PATH = Path.home()
+DATASET_PATH = HOME_PATH / ".cache/nncf/datasets/mvtec"
+CHECKPOINT_PATH = HOME_PATH / ".cache/nncf/models/stfpm_mvtec"
+ROOT = Path(__file__).parent.resolve()
+FP32_RESULTS_ROOT = ROOT / "fp32"
+INT8_RESULTS_ROOT = ROOT / "int8"
+CHECKPOINT_URL = "https://huggingface.co/alexsu52/stfpm_mvtec_capsule/resolve/main/qat/model.ckpt"
```
Comment (AlexKoff88) [inline examples/quantization_aware_training/torch/anomalib/main.py:34]: we found this is not a good practice to share stuff from a private account. We should revise the samples where there is such a behavior. The best alternative is to use OpenVINO S3.

File: examples/quantization_aware_training/torch/anomalib/README.md (hunk 5 above)
Comment (alexsu52) [reply examples/quantization_aware_training/torch/anomalib/README.md:5]: done

File: examples/quantization_aware_training/torch/anomalib/README.md (hunk 4 above)
Comment (alexsu52) [reply examples/quantization_aware_training/torch/anomalib/README.md:7]: done

File: examples/quantization_aware_training/torch/anomalib/README.md (hunk 6 above)
Comment (alexsu52) [reply examples/quantization_aware_training/torch/anomalib/README.md:15]: done

File: examples/quantization_aware_training/torch/anomalib/main.py (hunk 8 above)
Comment (alexsu52) [reply examples/quantization_aware_training/torch/anomalib/main.py]: done

**Pull request merged**